src/coreclr/src/vm/threadsuspend.cpp

```diff
@@ -3896,138 +3799,19 @@ HRESULT ThreadSuspend::SuspendRuntime(ThreadSuspend::SUSPEND_REASON reason)
-
-                if (str == Thread::STR_Success)
-                    g_SuspendStatistics.cntOSSuspendResume++;
-                else
-                    g_SuspendStatistics.cntFailedSuspends++;
-#endif
-
-#if defined(FEATURE_HIJACK) && !defined(TARGET_UNIX)
-                // Only check HandledJITCase if we actually suspended the thread, and
-                // the thread is in cooperative mode.
-                // See comment at the previous invocation of HandledJITCase - it does
-                // more than you think!
-                if (str == Thread::STR_Success && thread->m_fPreemptiveGCDisabled)
-                {
-                    Thread::WorkingOnThreadContextHolder workingOnThreadContext(thread);
-                    if (workingOnThreadContext.Acquired() && thread->HandledJITCase())
-                    {
-                        // Redirect thread so we can capture a good thread context
-                        // (GetThreadContext is not sufficient, due to an OS bug).
-                        if (!thread->CheckForAndDoRedirectForGC())
-                        {
-#ifdef TIME_SUSPEND
-                            g_SuspendStatistics.cntFailedRedirections++;
-#endif
-                            STRESS_LOG1(LF_SYNC, LL_INFO1000, "Failed to CheckForAndDoRedirectForGC(). Retry suspension 2 for thread %p\n", thread);
-                            thread->ResumeThread();
-                            goto RetrySuspension2;
-                        }
-#ifdef TIME_SUSPEND
-                        else
-                            g_SuspendStatistics.cntRedirections++;
-#endif
-                    }
-                }
-#endif // FEATURE_HIJACK && !TARGET_UNIX
-
-                if (str == Thread::STR_Success)
-                    thread->ResumeThread();
-            }
-#endif // DISABLE_THREADSUSPEND
-        }
-        else
-        if (res == WAIT_OBJECT_0)
+       }
+        else if (res == WAIT_OBJECT_0)
         {
-            g_pGCSuspendEvent->Reset();
-            continue;
         }
         else
         {
             // No WAIT_FAILED, WAIT_ABANDONED, etc.
             _ASSERTE(!"unexpected wait termination during gc suspension");
         }
-    }
-
-#ifdef PROFILING_SUPPORTED
-    // If a profiler is keeping track of GC events, notify it
-    {
-        BEGIN_PIN_PROFILER(CORProfilerTrackSuspends());
-        g_profControlBlock.pProfInterface->RuntimeSuspendFinished();
-        END_PIN_PROFILER();
-    }
-#endif // PROFILING_SUPPORTED
 
-#ifdef _DEBUG
-    if (reason == ThreadSuspend::SUSPEND_FOR_GC) {
-        thread = NULL;
-        while ((thread = ThreadStore::GetThreadList(thread)) != NULL)
-        {
-            thread->DisableStressHeap();
-            _ASSERTE (!thread->HasThreadState(Thread::TS_GCSuspendPending));
-        }
+        observeOnly = false;
+        g_pGCSuspendEvent->Reset();
     }
-#endif
 
     // We know all threads are in preemptive mode, so go ahead and reset the event.
```

Comment:
>// We  [](start = 4, length = 6)

Looks like we are missing `g_profControlBlock.pProfInterface->RuntimeSuspendFinished();` here

Reply:
`thread->DisableStressHeap();`  loop should be here too

---
In reply to: [505926851](https://github.com/dotnet/runtime/pull/34666#discussion_r505926851) [](ancestors = 505926851)